{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove Gaps\n",
    "\n",
    "This script removes gaps in the reads and the genome. It contains the following steps:\n",
    "\n",
    "- Loading and cleaning read data and reference genome data\n",
    "- Visualizing overlap between reads\n",
    "- Collapsing the regions where there is empty data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input variables\n",
    "run_number=\"run3\"\n",
    "chrom=\"chr17\"\n",
    "dis=\"pompe\"\n",
    "# Setup\n",
    "chrom_dis=f\"{chrom}_{dis}\"\n",
    "rootdir=f\"/mnt/aretian/genomics/nanopore\"\n",
    "datadir=f\"/mnt/aretian/genomics/nanopore/{run_number}\"\n",
    "\n",
    "os.environ[\"run_number\"]=run_number\n",
    "os.environ[\"chrom\"]=chrom\n",
    "os.environ[\"chrom_dis\"]=chrom_dis\n",
    "os.environ[\"datadir\"]=datadir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and Clean Reference Genome\n",
    "\n",
    "Info: https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use\n",
    "\n",
    "S3 nanopore directory:\n",
    "\n",
    "s3cmd get --recursive s3://aretian-genomics/nanopore/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull reference genome from S3\n",
    "# !s3cmd get s3://aretian-genomics/nanopore/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz\n",
    "\n",
    "# !gunzip GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz\n",
    "\n",
    "# Index reference genome\n",
    "# !samtools faidx GCA_000001405.15_GRCh38_no_alt_analysis_set.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract chromosome from reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# echo \"Extracting $chrom from ${datadir}/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz\"\n",
    "# echo \"Writing to ${rootdir}/reference_genome_${chrom}.fa\"\n",
    "\n",
    "# /home/fer/miniconda3/envs/genomics/bin/samtools faidx \"${datadir}/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz\" $chrom > \"${rootdir}/reference_genome_${chrom}.fa\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique characters: ['A', 'W', 'S', 'C', 'G', 'T', 'N', 'K', 'R', 'Y']\n",
      "Selected chromosome from reference genome is 83257441 BP long\n"
     ]
    }
   ],
   "source": [
    "# Read in fasta file: remove line breaks and header\n",
    "def read_fasta_genome(fasta_file,chromosome_header):\n",
    "    clean_data = fasta_file.read().replace(\"\\n\", \"\")\n",
    "    clean_data = clean_data.replace(chromosome_header,\"\") # get rid of header\n",
    "\n",
    "    return clean_data\n",
    "\n",
    "with open(f'{rootdir}/reference_genome_{chrom}.fa') as f: # update path if needed\n",
    "    ref_genome = read_fasta_genome(f,f'>{chrom}')\n",
    "\n",
    "# See https://www.bioinformatics.org/sms/iupac.html for IUPAC nucleotide codes\n",
    "\n",
    "print(f\"Unique characters: {list(set(ref_genome))}\") \n",
    "\n",
    "print(f\"Selected chromosome from reference genome is {len(ref_genome)} BP long\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and Clean Reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Samtools view --> format of output: http://samtools.github.io/hts-specs/SAMv1.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read SAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "usecols_list = list(range(11))\n",
    "# nanopore_reads = pd.read_csv(f'{datadir}/{run_number}_{chrom_dis}.sam', sep='\\t',header=None,error_bad_lines=False, warn_bad_lines=False)\n",
    "nanopore_reads = pd.read_csv(f'{datadir}/{run_number}_{chrom_dis}.sam',usecols=usecols_list, sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "nanopore_reads = nanopore_reads.iloc[:,:11]\n",
    "nanopore_reads.columns = ['QNAME', 'FLAG', 'RNAME', 'POS', 'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT','TLEN', 'SEQ', 'QUAL']\n",
    "\n",
    "# Sort and get ID\n",
    "nanopore_reads = nanopore_reads.sort_values(by='POS',ascending=True) # sort based on starting index of reads\n",
    "nanopore_reads['ID'] = np.nan\n",
    "\n",
    "# Get columns of interest\n",
    "nanopore_reads = nanopore_reads[['ID', 'QNAME', 'FLAG', 'RNAME', 'POS', 'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT','TLEN', 'SEQ', 'QUAL']]\n",
    "\n",
    "# Drop missing values\n",
    "nanopore_reads = nanopore_reads.loc[nanopore_reads['SEQ'] != '*'] # drop any rows without a proper nucleotide sequence\n",
    "\n",
    "# Reset index and save ID\n",
    "nanopore_reads = nanopore_reads.reset_index()\n",
    "nanopore_reads['ID'] = nanopore_reads.index\n",
    "\n",
    "# Calculate END_POS\n",
    "nanopore_reads['SEQ_LEN'] = nanopore_reads['SEQ'].apply(lambda x: len(x))\n",
    "nanopore_reads['END_POS'] = nanopore_reads['POS']+nanopore_reads['SEQ_LEN']-1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Descriptive Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count and plot overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_reads_length():\n",
    "    for ID in nanopore_reads.index:\n",
    "        print(f\"Leftmost index --> {nanopore_reads.loc[ID,'POS']}, Read length --> {len(nanopore_reads.loc[ID,'SEQ'])}\")\n",
    "\n",
    "def create_overlap_df():\n",
    "    end_index = max(nanopore_reads['POS']) + len(nanopore_reads.loc[nanopore_reads['POS'].idxmax(),'SEQ'])\n",
    "    start_index = min(nanopore_reads['POS'])\n",
    "    index_range = end_index-start_index\n",
    "\n",
    "    read_overlap = np.zeros((len(nanopore_reads),index_range)) # empty matrix for reads\n",
    "\n",
    "    for ID in nanopore_reads.index:\n",
    "    #     print(f\"Adding raad cover for read {ID+1}: Index-->{nanopore_reads.loc[ID,'POS']}, length-->{len(nanopore_reads.loc[ID,'SEQ'])}\")\n",
    "        start = nanopore_reads.loc[ID,'POS']-start_index\n",
    "        end = start+len(nanopore_reads.loc[ID,'SEQ'])\n",
    "        read_overlap[ID,start:end] = 1\n",
    "\n",
    "    summary = read_overlap.sum(axis=0)\n",
    "    \n",
    "    overlap_df = pd.DataFrame(data=summary,index=range(start_index,end_index), columns=['overlap_count'])\n",
    "    overlap_df['position'] = overlap_df.index\n",
    "    overlap_df['isempty']=overlap_df['overlap_count']==0\n",
    "    \n",
    "    return overlap_df\n",
    "\n",
    "def count_empty_regions():\n",
    "    win_length=10000\n",
    "    emptycount = 0\n",
    "    emptyregioncounter = 0\n",
    "\n",
    "    for i in overlap_df.index:\n",
    "\n",
    "        if overlap_df.loc[i,'isempty']==False:\n",
    "            emptycount = 0\n",
    "            continue\n",
    "\n",
    "        emptycount += 1\n",
    "\n",
    "        if emptycount >= win_length:\n",
    "            emptyregioncounter +=1\n",
    "\n",
    "        if i%100000==0:\n",
    "            print(f\"Loop: {i}\")\n",
    "\n",
    "    print(f\"There are at most {emptyregioncounter} regions with lengths larger than {win_length}\")\n",
    "\n",
    "def plot_overlap(overlap_df):\n",
    "    fig, ax = plt.subplots(dpi=150)\n",
    "    # plt.plot(overlap_df[:40000]['position'],overlap_df[:40000]['overlap_count'])\n",
    "    # plt.plot(overlap_df[:40000]['position'],overlap_df[:40000]['overlap_count'])\n",
    "    # plt.plot(overlap_df.loc[1000000:,'position'],overlap_df.loc[1000000:,'overlap_count'])\n",
    "    # plt.plot(overlap_df['position'],overlap_df['overlap_count'])\n",
    "    plt.plot(overlap_df['position'],overlap_df['isempty'])\n",
    "    plt.xlabel('Nucleotide Position')\n",
    "    plt.ylabel('# Reads Overlapping')\n",
    "\n",
    "# count_empty_regions()\n",
    "# overlap_df = create_overlap_df()\n",
    "# plot_overlap(overlap_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read Length and Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read length statistics:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count      1161.000000\n",
       "mean      14356.910422\n",
       "std       18321.121713\n",
       "min         222.000000\n",
       "25%        2156.000000\n",
       "50%        7406.000000\n",
       "75%       18708.000000\n",
       "max      130186.000000\n",
       "Name: SEQ_LEN, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check read length statistics\n",
    "print(f\"Read length statistics:\")\n",
    "nanopore_reads['SEQ_LEN'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read position statistics:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    1.161000e+03\n",
       "mean     5.256842e+06\n",
       "std      1.138441e+06\n",
       "min      3.199746e+06\n",
       "25%      4.278872e+06\n",
       "50%      5.255851e+06\n",
       "75%      6.216842e+06\n",
       "max      7.218912e+06\n",
       "Name: POS, dtype: float64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check read pos statistics\n",
    "print(f\"Read position statistics:\")\n",
    "nanopore_reads['POS'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim reference genome chromosome to locus of interest "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collapse Gaps "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract location of interest from Reference Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected locus for chromosome from reference genome is 4080269 BP long\n"
     ]
    }
   ],
   "source": [
    "ref_genome = ref_genome[min(nanopore_reads['POS']):max(nanopore_reads['END_POS'])]\n",
    "print(f\"Selected locus for chromosome from reference genome is {len(ref_genome)} BP long\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reset start and end positions based on location of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "shift = min(nanopore_reads['POS'])\n",
    "\n",
    "# archive old positions\n",
    "nanopore_reads['ORIG_POS'] = nanopore_reads['POS']\n",
    "nanopore_reads['ORIG_END_POS'] = nanopore_reads['END_POS']\n",
    "\n",
    "# shift positions\n",
    "nanopore_reads['POS'] = nanopore_reads['POS']-shift\n",
    "nanopore_reads['END_POS'] = nanopore_reads['END_POS']-shift\n",
    "\n",
    "nanopore_reads = nanopore_reads.reset_index() # to make the next step easier\n",
    "nanopore_reads = nanopore_reads[['ID', 'QNAME', 'FLAG', 'RNAME', 'POS', 'MAPQ', 'CIGAR','RNEXT', 'PNEXT', 'TLEN', 'SEQ', 'QUAL', 'SEQ_LEN', 'END_POS','ORIG_POS', 'ORIG_END_POS']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 29 gaps.\n"
     ]
    }
   ],
   "source": [
    "confirmed_gaps = [] # Collect gaps in list\n",
    "max_end_pos = 0 # Store end position of longest read\n",
    "# Initialize island counter\n",
    "islandID_counter=0\n",
    "nanopore_reads.loc[index:,\"islandID\"]=islandID_counter            \n",
    "\n",
    "for index in nanopore_reads.index:\n",
    " \n",
    "    current_read_end = nanopore_reads.loc[index,'END_POS']\n",
    "    \n",
    "    # Current read is redundant, skip\n",
    "    if current_read_end <= max_end_pos:\n",
    "        continue\n",
    "        \n",
    "    # This is a \"standard\" read (not embedded)    \n",
    "    elif current_read_end > max_end_pos: \n",
    " \n",
    "        current_read_start = nanopore_reads.loc[index,'POS']\n",
    "        \n",
    "        # There is no gap, update max_end_pos\n",
    "        if current_read_start <= max_end_pos: \n",
    "            max_end_pos = current_read_end \n",
    "            continue\n",
    "            \n",
    "        # There is a gap, remove it  \n",
    "        elif current_read_start > max_end_pos: \n",
    "            gap_start = max_end_pos+1\n",
    "            gap_end = current_read_start-1\n",
    "            shift = gap_end - gap_start + 1\n",
    "         \n",
    "            nanopore_reads.loc[index:,'POS']     = nanopore_reads.loc[index:,'POS']    -shift\n",
    "            nanopore_reads.loc[index:,'END_POS'] = nanopore_reads.loc[index:,'END_POS']-shift\n",
    "\n",
    "            # Calculate gap based on original position\n",
    "            orig_gap_end   = nanopore_reads.loc[index,'ORIG_POS'] - 1\n",
    "            orig_gap_start = orig_gap_end - shift + 1\n",
    "            orig_gap = (orig_gap_start,orig_gap_end)\n",
    "            confirmed_gaps.append(orig_gap) # Add gap to list\n",
    "     \n",
    "            current_read_end = nanopore_reads.loc[index,'END_POS']\n",
    "            max_end_pos = current_read_end # Current read end should always be larger than max_end_pos\n",
    "            \n",
    "            islandID_counter += 1\n",
    "            nanopore_reads.loc[index:,\"islandID\"]=islandID_counter            \n",
    "            \n",
    "#             nanopore_reads.loc[:index,\"islandID\"]=islandID_counter\n",
    "    \n",
    "\n",
    "#      try:       \n",
    "#         # Current read is not redundant, then update max_end_pos\n",
    "#         max_end_pos = current_read_end\n",
    "        \n",
    "#         next_read_start = nanopore_reads.loc[index+1,'POS']\n",
    "\n",
    "#         delta = next_read_start-current_read_end\n",
    "\n",
    "#         if delta <= 0: # There is overlap, so no problem\n",
    "#             continue \n",
    "#         elif delta > 0: # There is a gap\n",
    "#             # Calculate gap based on original position\n",
    "#             gap_start = nanopore_reads.loc[index  ,'ORIG_END_POS']\n",
    "#             gap_end   = nanopore_reads.loc[index+1,'ORIG_POS']\n",
    "#             orig_gap = (gap_start+1,gap_end-1)\n",
    "#             confirmed_gaps.append(orig_gap) # Add gap to list\n",
    "            \n",
    "#             # Calculate gap based on adjusted positions\n",
    "#             gap = (current_read_end+1,next_read_start-1)\n",
    "          \n",
    "#             # Shift remaining reads to the left\n",
    "#             shift = gap[1]-gap[0]+1 \n",
    "#             # Check shift is the same with original position data\n",
    "#             shift_check = orig_gap[1]-orig_gap[0]+1\n",
    "#             if shift!=shift_check:\n",
    "#                 print(f\"Found difference: orig_gap: {orig_gap}, gap: {gap}.\")\n",
    "#                 print(f\"Difference: orig_gap {shift_check}, gap {shift}\")\n",
    "#             try:\n",
    "#                 nanopore_reads.loc[(index+1):,'POS'] = nanopore_reads.loc[index+1:,'POS']-shift\n",
    "#                 nanopore_reads.loc[(index+1):,'END_POS'] = nanopore_reads.loc[index+1:,'END_POS']-shift\n",
    "#             except:\n",
    "#                 pass\n",
    "\n",
    "#     except:\n",
    "#         pass\n",
    "            \n",
    "print(f\"There are {len(confirmed_gaps)} gaps.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check if gaps overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "previous_gap_end = 0\n",
    "for gap in confirmed_gaps:\n",
    "    gap_start = gap[0]\n",
    "    \n",
    "    if gap_start <= previous_gap_end: # we have a problem\n",
    "        print(f\"Found an intersection in gap {gap}.\")\n",
    "    \n",
    "    # Update gap_end for next iteration\n",
    "    previous_gap_end = gap[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot shift in gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzIAAAI/CAYAAACs3OxHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAsi0lEQVR4nO3df7RlZX0n6M8XKKy0QphAtRKgLBToqJEoUyoEE20laURau6MucZaJcXVWjUZnkqxkzbLTmaSTOKvTcdr0cjAQkhghyZB02kjoNKZNOj+dAREBEVS6y4RoBaIGRoQgIvjOH/dUWZa36p5779737L3P86x14PzYd5+33vfsc87nvO+732qtBQAAYEyOWnQBAAAA1kuQAQAARkeQAQAARkeQAQAARkeQAQAARkeQAQAARueYRT3xSSed1Hbt2rWopwcAAAbuwx/+8N+11nas9tjCgsyuXbty0003LerpAQCAgauqvz7cY4aWAQAAoyPIAAAAoyPIAAAAo7OwOTIAAMB8vvzlL2ffvn15+OGHF12UXmzfvj2nnnpqtm3bNvffCDIAADBw+/bty3HHHZddu3alqhZdnE611nLvvfdm3759Of300+f+O0PLAABg4B5++OGceOKJkwsxSVJVOfHEE9fd2yTIAADACEwxxOy3kX+bIAMAAIyOIAMAAHTiscce27LnEmQAAIA13XXXXfmWb/mWvO51r8vZZ5+dV77ylXnooYeya9eu/MzP/Eye//zn53d+53fy/ve/P+edd17OOeecvOpVr8qDDz7YS3kEGQAAYC533nln9uzZk9tuuy3HH398fvEXfzHJyumTP/CBD+SCCy7IW9/61vzRH/1Rbr755uzevTtvf/vbeymL0y8DAMCI/PR/uiMfu/sLne7z6d98fH7qnz5jze1OO+20nH/++UmS1772tXnHO96RJHn1q1+dJLnhhhvysY997MA2jzzySM4777xOy7qfIAMAAMzl0LOL7b/9+Mc/PsnKmjDf9V3flauvvrr3sggyAAAwIvP0nPTlU5/6VK6//vqcd955ufrqq/P85z8/t9xyy4HHzz333LzpTW/K3r17c8YZZ+Shhx7Kvn37ctZZZ3VeFnNkAACAuTztaU/LlVdembPPPjv33Xdf3vjGN37N4zt27Mi73/3uvOY1r8nZZ5+dc889N5/4xCd6KYseGQAAYC5HHXVULr/88q+576677vqa2y960YvyoQ99qP+y9P4MAAAAHRNkAACANe3atSu33377ootxgCADAACMztxBpqqOrqpbqur3V3msquodVbW3qm6rqnO6LSYAAMBXradH5oeSfPwwj70kyZmzy54kl22yXAAAAIc1V5CpqlOTvDTJrxxmk5cnuaqtuCHJCVV1ckdlBAAA+Brznn753yf535Icd5jHT0ny6YNu75vdd8+GSwawTL7yWPKbr0zu39fP/r/jx5Jve3U/+waAOV1zzTU566yz8vSnP33T+1ozyFTVxUk+21r7cFW98HCbrXJfW2Vfe7Iy9Cw7d+6cv5QAU/fw/ckn/zh50jOTE8/odt93/kHyV38uyACwcNdcc00uvvjirQkySc5P8rKquijJ9iTHV9VvtNZee9A2+5KcdtDtU5PcfeiOWmtXJLkiSXbv3v11QQdg6T3rtcm5b+h2n29/Rrf7A2Ap3XXXXbnwwgvzvOc9L7fcckvOOuusXHXVVbn++uvzYz/2Y3n00UfznOc8J5dddlke97jH5S1veUuuvfbaHHPMMfnu7/7ufM/3fE+uvfba/Nmf/Vne+ta35j3veU+e+tSnbrg8a86Raa39y9baqa21XUkuSfLHh4SYJLk2yffNzl52bpL7W2uGlQEAwITceeed2bNnT2677bYcf/zxefvb357v//7vz2//9m/nox/9aB599NFcdtllue+++/Le9743d9xxR2677bb8xE/8RL792789L3vZy/K2t70tt95666ZCTDL/HJmvU1VvSJLW2uVJrktyUZK9SR5K8vpNlQoAAFjd+96S/O1Hu93nk56ZvOTn1tzstNNOy/nnn58kee1rX5uf/dmfzemnn56zzjorSfK6170u73znO/PmN78527dvzw/8wA/kpS99aS6++OJuy5t1LojZWvvT1trFs+uXz0JMZmcre1Nr7amttWe21m7qvKQAbILRvABsXtVqU+O/3jHHHJMbb7wxr3jFK3LNNdfkwgsv7LwsG+6RAaAHc35ALHyfACzOHD0nffnUpz6V66+/Puedd16uvvrqXHDBBfmlX/ql7N27N2eccUZ+/dd/PS94wQvy4IMP5qGHHspFF12Uc889N2ecsXIim+OOOy4PPPBAJ2VZV48MAACwvJ72tKflyiuvzNlnn5377rsvP/IjP5Jf+7Vfy6te9ao885nPzFFHHZU3vOENeeCBB3LxxRfn7LPPzgte8IL8wi/8QpLkkksuydve9rY8+9nPzic/+clNlUWPDAAAMJejjjoql19++dfc9+IXvzi33HLL19x38skn58Ybb/y6vz///PPzsY99rJuydLIXADan9TyHpe/9A8AWE2QABsV8FgCGadeuXbn99tsXXYwDBBmAyROOAJgeQQYAAEagTXiY8Eb+bYIMwCD0/eE03Q8/gGWwffv23HvvvZMMM6213Hvvvdm+ffu6/s5ZywCGxJovAKzi1FNPzb59+/K5z31u0UXpxfbt23Pqqaeu628EGYCpk40ARm/btm05/fTTF12MQTG0DAAAGB1BBmAIrCMDAOsiyAAAAKMjyABMnkkyAEyPIAMAAIyOIAMwCOawAMB6CDIAS0FQAmBaBBmAIeljQUyLbAIwQYIMAAAwOoIMwBBY5wUA1kWQAVgGghIAEyPIAAyK+SwAMA9BBmDyhCMApkeQARgEQ78AYD0EGYClICgBMC2CDMCQWPMFAOYiyABMnXAEwAQJMgBD4PTIALAuggzAMhCUAJgYQQZgUAwDA4B5CDIAkyccATA9ggzAIBj6BQDrIcgAAACjI8gADElvp0rW4wPAtAgyAFNnHRkAJkiQARgCp0cGgHURZAAAgNERZAAGpadhYHp8AJgYQQYAABgdQQZgEPrsMTHZH4DpEWQAAIDREWQAhsQ6MgAwF0EGAAAYHUEGYOosiAnABAkyAEPg9MgAsC6CDMCgWEcGAOYhyAAAAKMjyABMnjkyAEyPIAMwCIZ+AcB6CDIAQ2IdGQCYiyADAACMjiADMHXWkQFgggQZgCFwemQAWJc1g0xVba+qG6vqI1V1R1X99CrbvLCq7q+qW2eXn+ynuABTp/cEAOZxzBzbfCnJi1prD1bVtiQfqKr3tdZuOGS7v2itXdx9EQHYND0+AEzMmkGmtdaSPDi7uW128YkIMBp6eQCYnrnmyFTV0VV1a5LPJvnD1toHV9nsvNnws/dV1TO6LCTA9Pl9CADWY64g01p7rLX2rCSnJnluVX3rIZvcnOTJrbVvS/J/Jblmtf1U1Z6quqmqbvrc5z638VIDAABLbV1nLWutfT7Jnya58JD7v9Bae3B2/bok26rqpFX+/orW2u7W2u4dO3ZsuNAAk2VBTACYyzxnLdtRVSfMrn9DkguSfOKQbZ5UtfLpW1XPne333s5LCwAAkPnOWnZykiur6uisBJT/0Fr7/ap6Q5K01i5P8sokb6yqR5N8Mckls5MEADCPPt8yLYgJwATNc9ay25I8e5X7Lz/o+qVJLu22aAAAAKtb1xwZAPrWU++JTnIAJkaQAQAARkeQARiEPntMzJEBYHoEGQAAYHQEGYAhcYYxAJiLIAMAAIyOIAMwBNaRAYB1EWQAAIDREWQABkXvCQDMQ5ABWAYWxARgYgQZgMnTywPA9AgyAADA6AgyAEPiDGMAMBdBBmApmCMDwLQIMgBDYDI+AKyLIAMwdUarATBBggzAoEgdADAPQQZgGRi6BsDECDIAgyBoAMB6CDIAk2e4GgDTI8gADIl1ZABgLoIMwFIwdA2AaRFkAIbAZHwAWBdBBmDqDFcDYIIEGQAAYHQEGQAAYHQEGYBlYA4OABMjyAAMQp9BwxwZAKZHkAEYEhPzAWAuggwAADA6ggzAUjBHBoBpEWQAhsBkfABYF0EGYFB6mCNj3g0AEyTIAAAAoyPIACwDQ9cAmBhBBmAQBA0AWA9BBmBIepnPYo4MANMjyAAAAKMjyAAsBUPXAJgWQQZgCEzGB4B1EWQAps46MgBMkCADMChCBwDMQ5ABAABGR5ABGISe58iYgwPAxAgyAJNnuBoA0yPIAAyJifkAMBdBBgAAGB1BBmAIep/DYo4MANMiyABMneFqAEyQIAMwKEIHAMxDkAEAAEZHkAEYBOvIAMB6rBlkqmp7Vd1YVR+pqjuq6qdX2aaq6h1Vtbeqbquqc/opLgAAQHLMHNt8KcmLWmsPVtW2JB+oqve11m44aJuXJDlzdnlekstm/wdgPXqZmG/eDQDTs2aPTFvx4Ozmttnl0DEKL09y1WzbG5KcUFUnd1tUAACAFfP0yKSqjk7y4SRnJHlna+2Dh2xySpJPH3R73+y+e7ooJCPz8P3Jp27odkz+SWcmJz61u/3BIt19a/LA337tff/fXf0+59//XXLnH6z/7xx7AAzUXEGmtfZYkmdV1QlJ3ltV39pau/2gTVYbt/B132Krak+SPUmyc+fO9ZeWcfizn0+uv7TbfZ50VvLmD3W7T1iER/4++eUXJe2x1R9/3PHdP+f2b0z23Zhc/er1/+2JZyT/y4e7LxMAbNJcQWa/1trnq+pPk1yY5OAgsy/JaQfdPjXJ3av8/RVJrkiS3bt3O4XOVD3y98n2E5LvfW83+/uzn0/+9qPd7AsW7dEvrYSY896cfOsrvvaxbd+Q7PiW7p/zle9K7t27/r/78/8zufuW7ssDAB1YM8hU1Y4kX56FmG9IckGSf3vIZtcmeXNV/VZWJvnf31ozrGyZHX1sckpHJ6/7Byd2sx8Ykm88rbtjZC3bj9/Ycz3esQfAcM3TI3Nykitn82SOSvIfWmu/X1VvSJLW2uVJrktyUZK9SR5K8vqeygsAALB2kGmt3Zbk2avcf/lB11uSN3VbNMarj1GDRiLCYjj2ABimNU+/DAAAMDSCDP3oclE/a/kxRb0sfNm1MZQRgGUlyAAAAKMjyNC9LhfC7HOfsAhjey2PrbwALA1BBgAAGB1Bhp50ObbeOH2maASv61HM4wFgWQkyAADA6Agy9MA6MjAdjj0AhkmQAdhSggEAdEGQoR+driNjnD4TNIrX9RjKCMCyEmQAAIDREWTonnVkYDocewAMlCADsJUEAwDohCBDT6wjAwBAfwQZAFY3ihMSALCsBBl6YB0ZmA7HHgDDJMgAbCnBAAC6IMjQD+vIwJF5XQPApggyAByGsAXAcAkyAADA6AgydK+Xuf7mFTARY3stj628ACwNQYaeWEcGjszrGgA2Q5ABYHVOSADAgAkyAADA6Agy9MCCmDAdjj0AhkmQoR/WkYHDmAUDr2sA2BRBBoDDELYAGC5BBgAAGB1Bhu71se6EtSxgMRx7AAyUIENPrCMDqzoQDLyuAWAzBBkAAGB0BBkAVufMagAMmCBDD6wjA9Ph2ANgmAQZgC0lGABAFwQZ+tHpXH/DW5ggr2sA2BRBBoDDELYAGC5Bhu5ZRwamw7EHwEAJMgBbSTAAgE4IMvTEgphwZF7XALAZggwAq3NCAgAGTJChB9aRgelw7AEwTIIMAAAwOoIM/ehySIrhLUzKrIfD6xoANkWQAeAwhC0AhkuQoXvWkYHpcOgBMFCCDAAAMDqCDD2xjgys6kDvotc1AGyGIAMAAIyOIEMPrCMDk1AVxx4AQyXIAAAAoyPI0A/ryMBhWEcGALogyAAAAKMjyNC9XtaR6X6XwFrKGk4ADNaaQaaqTquqP6mqj1fVHVX1Q6ts88Kqur+qbp1dfrKf4gIAACTHzLHNo0l+tLV2c1Udl+TDVfWHrbWPHbLdX7TWLu6+iIyTdWRgVdaRAYBOrNkj01q7p7V28+z6A0k+nuSUvgsGAABwOOuaI1NVu5I8O8kHV3n4vKr6SFW9r6qe0UXhGCvryMAkWEcGgAGbZ2hZkqSqnpDkPUl+uLX2hUMevjnJk1trD1bVRUmuSXLmKvvYk2RPkuzcuXOjZQYAAJbcXD0yVbUtKyHmN1trv3vo4621L7TWHpxdvy7Jtqo6aZXtrmit7W6t7d6xY8cmi86gWUcGDsM6MgDQhXnOWlZJfjXJx1trbz/MNk+abZeqeu5sv/d2WVAAAID95hladn6S703y0aq6dXbfjyfZmSSttcuTvDLJG6vq0SRfTHJJaxYfWFq9rCPj5QQL4dgDYKDWDDKttQ9kjfOEttYuTXJpV4UCAAA4knWdtQzmZx0ZODKvawDYDEEGYCsZqgUAnRBkAFidM6sBMGCCDD2wICZMh2MPgGESZOiHdWTgyLyuAWBTBBmALaWHAwC6IMgAAACjI8jQPQtiwkSUYw+AwRJk6Inx/3BkjhEA2AxBBmAr6eEAgE4IMgAAwOgIMvTAOjIwCVVx7AEwVIIM/bCODByZ1zUAbIogAwAAjI4gAwAAjI4gQ/esIwMTYR0ZAIZLkAEAAEZHkKEnXU5kNimaKfK6BoDNEGQAtpKhWgDQCUGGHlhHBibBOjIADJggAwAAjI4gQz8siAlH5nUNAJsiyABsKUO1AKALggzds44MTIdjD4CBEmQAOAzD3wAYLkGGnlhHBgCA/ggyAFvJUC0A6IQgw0j48geL4dgDYJgEGQBW5xTRAAyYIEM/rCMDR+Z1DQCbIsgAbClDtQCgC4IM3bOODEyHYw+AgRJkADgMw98AGC5Bhp5YRwaOzOsaADZDkAEAAEZHkKEHfYypN06fiRjdnJOxlReAZSHIALA6p4gGYMAEGfrR6RQZX6aYIK9rANgUQQYAABgdQYbuWUcGjmBkr2XHHgADJcgAAACjI8jQE+vIwJGN4XU9hjICsKwEGQAAYHQEGXpgHRk4rNHNORlbeQFYFoIMAAAwOoIM/ehyjQzrbTBFY3hdj6GMACwtQQYAABgdQQZgS5lzAgBdEGTongUxAQDomSBDT6wjA0c2htf1GMoIwLISZAAAgNERZAAAgNERZOiBBTHhsMY432uMZQZg8tYMMlV1WlX9SVV9vKruqKofWmWbqqp3VNXeqrqtqs7pp7iMhnVk4MjG8LoeQxkBWFrHzLHNo0l+tLV2c1Udl+TDVfWHrbWPHbTNS5KcObs8L8lls/8DAAB0bs0emdbaPa21m2fXH0jy8SSnHLLZy5Nc1VbckOSEqjq589ICAABkvh6ZA6pqV5JnJ/ngIQ+dkuTTB93eN7vvns0Ubkt8+eHkN16x6FJMy2duT77p9O73+2sv7X6fffnmZyX/5P9YdCnYCn9/b/J7P5h86cH5tn9kzu2G5N0vzeBOxfxtlyTnfO+iSzEu9/1V8p9/NHn0S4suydc7elty4c8l//BbFl0SYETmnuxfVU9I8p4kP9xa+8KhD6/yJ183O7Sq9lTVTVV10+c+97n1lZTxeOK3Jme/urv9PfXFyenf2d3++nbfXyYfvnLRpWCrfPaO5L/9QfLw/fNtf+wTkjMuSL752f2WqwsHjr2BhZh7PpLc8buLLsX47Lsp+eR/Tb780KJL8rUe+1Lyl3+SfOr/XXRJgJGZq0emqrZlJcT8ZmtttU+PfUlOO+j2qUnuPnSj1toVSa5Ikt27dw/jNDjbtiev/8+LLgVHsvN5yev+06JLMb8/+PHk5qsWXQq22kt+Ltn1/EWXolunPWeYx96vXLDoEozb9/xyctIZiy7FVz3wt8m/+0eLLgUwQvOctayS/GqSj7fW3n6Yza5N8n2zs5edm+T+1trwh5UBAACjNE+PzPlJvjfJR6vq1tl9P55kZ5K01i5Pcl2Si5LsTfJQktd3XlIAAICZNYNMa+0DWWOAdGutJXlTV4WC8RvGyEm2gMUiF0O9b8DA60ybAus092R/YE4WEVxS2n3rqOtNGdx71NDKA4yFIAMAAIyOIAMAAIyOIAN9MNZ7iWjrxVDv6zb496Whlw8YGkEGoAuDm3cwYeoagAgyAMAiCabABgkyAADA6Agy0AtjvZfG4OcdTJR634CB15k2BdZJkAEAAEZHkIGuGe+9pLT71lHXmzK496ihlQcYC0EGAAAYHUEG+mCs9xLR1ouh3tfN+xIwMYIMAAAwOoIMdM5476U0uHkHE6auN2lg9ac9gQ0SZAAAgNERZKAXxqIvDfMOFkO9b4A6A6ZFkAEAFk84BdZJkIGuGe+9pLT71lHXmzK496ihlQcYC0EGAAAYHUEG+mCIxBLR1oyE9yVgYgQZAGAABC1gfQQZ6Jzx3ktpcPMOJkxdb5L6A6ZBkAEAFkcwBTZIkIFeGCKxNDT1YpjvsQHqDJgWQQYAWDzhFFgnQQa6ZpjEktLuW0ddb4r3KGAiBBkAAGB0BBnogyESS0RbL4Z6XzfvS8DECDIAwAAIWsD6CDLQOePPl5J5B1tHXW+S+gOmQZABABZHMAU2SJCBXhgisTTMO1gM9b4B6gyYFkEGAAAYHUEGumaYxJLS7ozEUN+j9LIB6yTIAAALNNBgBQyeIAN98MviEtHWi6He1837EjAxggwAADA6ggx0zjCJpaTZt85Q53iMxlDrT48RsD6CDAAAMDqCDACwOHrYgA0SZKAXhkgsDROoF0O9b4A6A6ZFkIGu+XVxSWn3raOuN2Wo71HCKbBOggwAADA6ggwAsEAD7SECBk+QgT4YIrFEtPViqPd1874ETIwgA53z6+JSGuq8gylS15s01PoTtID1EWQAAIDREWQAgMXRwwZskCADvTBEYmmYd7AY6n0D1BkwLYIMdM2vi0tKu28ddb0p3qOAiRBkAIDF08sGrJMgAwAskB4iYGMEGYBN8SvyYqj3ddPjAUzMmkGmqt5VVZ+tqtsP8/gLq+r+qrp1dvnJ7osJAADwVcfMsc27k1ya5KojbPMXrbWLOykRjJ5hEkvJBOqto643aaj1p8cIWJ81e2Raa3+e5L4tKAsAAMBcupojc15VfaSq3ldVz+honzBuxqMvB+28GOp9AwZaZ3rYgA2aZ2jZWm5O8uTW2oNVdVGSa5KcudqGVbUnyZ4k2blzZwdPDQAALKNN98i01r7QWntwdv26JNuq6qTDbHtFa213a233jh07NvvUMEx+XVxS2n3rqOtNGep7lF42YJ02HWSq6klVK++KVfXc2T7v3ex+AQAADmfNoWVVdXWSFyY5qar2JfmpJNuSpLV2eZJXJnljVT2a5ItJLmnNzyqQ1ob7yycd8na3GOp93Qb70ex9EtiYNYNMa+01azx+aVZOzwwAALAlujprGXCAXxeXkt63raOuN2mo9TfUHiNgqAQZAABgdAQZ6I1fF5fCYOcdTJx6nw49bMAGCTIAAMDoCDLQNb8uLintvnXU9aYM9T1KLxuwToIMAAAwOoIM9MWvi0tCOy+Gel+3wb4nDbSHCBg8QQYAABgdQQY659fFpTTUeQdTpK43Sf0B0yDIAAADMNShb8BQCTLQGx/KS2Gw8w4mTr1vwEDrTA8bsEGCDAAAMDqCDHTNj4tLSsNvHXW9KXpAgIkQZACAxTNcEFgnQQb64kN5SWjnxVDv6+Y9CZgYQQYAWCBD3YCNEWSgcz6Ul5J5B1tHXQMQQQYAGARD34D1EWSgNz6Ul4J5B4uh3jdAnQHTIsgAAItjqCCwQYIMdM2H8pLS7ltHXW+K9yhgIgQZAGDxjHwD1kmQAQAARkeQgb6YjLwktPNiqPd1G+x7kqFuwMYIMtA5H8pLybyDraOuN0n9AdMgyAAAAKMjyAAAAzDUoW/AUAky0BsfykthsPMOJk61b8BAK81QQWCDBBnomg/lJaXdt4663hTvUcBECDIAAMDoCDIAwOIZpgmskyADffGhDD1yfK2b9yRgYgQZ6Jzx50vJvIOto643aWj1N7TyAGMhyAAAAKMjyAAAA2DoG7A+ggz0xofyUjDvYDHU+waoM2BaBBkAYHHMeQI2SJCBrvlQXlLanZHwHgVMhCADACye4YLAOgky0BcfyktCOy+Gel8370nAxAgyAMDiGOoGbJAgA53zobyUfBnbOup6k9QfMA2CDAAAMDqCDPTGePSlYN7BYqj3DRh6nQ29fMDQCDIAAMDoCDLQNeP3oWeOsU3xHgVMhCADAACMjiADfTGGf0lo58VQ7+s29PekoZcPGBxBBgAAGB1BBjpn/PlSMu9g66jrTRpi/Q2xTMDQCTIAAMDoCDLQG+O9l4Jx/Yuh3jdg6HU29PIBQyPIAAAAo7NmkKmqd1XVZ6vq9sM8XlX1jqraW1W3VdU53RcTRsT4/SWl3beOut6UIb5HDbFMwODN0yPz7iQXHuHxlyQ5c3bZk+SyzRcLAADg8I5Za4PW2p9X1a4jbPLyJFe11lqSG6rqhKo6ubV2T1eF7NNjX2m54+77e32Os554XLZvO7rX52CA7vlIcuwTFl0K+vb5v150CZbTo19M/ubmRZdiXL5w96JLcGQP3KNNYU73PtzyN8c+pbfezBO+4djsPPEf9LLvLq0ZZOZwSpJPH3R73+y+UQSZRx79Sl526f/T63Nc8pzT8nOvOLvX52BAts0O/Cv/6WLLwdbaNvw3/Mk49vHJ5z+V/PI/XnRJxufoY5Ma4A9r2x6f3PIbKxdgTScm+d8f+V9z3VfO7WX//+xZ35x/f8mze9l3l7oIMqtFwVVPPVJVe7Iy/Cw7d+7s4Kk379hjjsqvvm53b/v/iWtuzwMPP9rb/hmgZ/1PyQmnJY9p96Xx+JOS45646FIsj+/62eRpL1t0Kcbp+G9Ojjl20aX4ev/i/SvhFFjbQ/cmv/eD+fZTt+UV/7if77BPPH57L/vtWhdBZl+S0w66fWqSVfuvW2tXJLkiSXbv3j2I8ywefVTlxU/r7wvIv3nfJ3rbNwN1zOOSMy5YdClguh5/YvKPjjR1k9F54tNXLsDa7v+bJMlJT3hcr99hx6CL0y9fm+T7ZmcvOzfJ/WOZHwMAAIzTmj0yVXV1khcmOamq9iX5qSTbkqS1dnmS65JclGRvkoeSvL6vwo5Vs8gXAABdsjDwXGcte80aj7ckb+qsRAAAAGvoYmgZR2CJLwAAOjM75bJ1ZAUZAABghASZLWAIIwAAdEuQAQCA0fFLuSDTM+MXAQDozmyOjJnYggwAADA+gswWMEcGAAC6JcgAAMDo+KVckOmZ8YsAAHRm/zoyCy7GEAgyAADA6AgyW6Dp+gMAgE4JMgAAMDp+KBdkemYdGQAA6J4gAwAAo2Gy/36CDAAAMDqCzBawICYAAN3yBVOQAQAARkeQAQCAsThwJimzZAQZAABgdASZLWAEIwAAXSrfMAWZvpWFZAAAoHOCDAAAjIYfyfcTZAAAgNERZLaAdWQAAKBbgkzPdP4BANCVNpvkb7K/IAMAACPiZ/L9BBkAAGB0BJktoesPAAC6JMgAAMBI7D+JlAFmgkzvrIcJAADdE2QAAGAs/Ep+gCCzBawjAwAA3RJkAABgJPb/Pm4dGUGmd3r/AACge4IMAACMhl/J9xNktoCOPwAA6JYgAwAAI9FmZ5HSLyPI9K68zAAAoHOCDAAAjIUzSR0gyGyBZiEZAADolCADAACMjiDTM71/AAB0xYKYXyXIAADAaNTX/G+ZCTJbQF4GAIBuCTIAAMDoCDI90+sHAEBX9p8Mt5wVV5ABAADGR5DZAgIzAACd2H9KXKfGFWQAAIDxEWT6Ji0DANCRNjsfrnVkBBkAAGCEBJktIC8DANANo332myvIVNWFVXVnVe2tqres8vgLq+r+qrp1dvnJ7osKAACw4pi1Nqiqo5O8M8l3JdmX5ENVdW1r7WOHbPoXrbWLeyjjqMnMAAB05cDZcJ0Wd64emecm2dta+8vW2iNJfivJy/stFgAAwOHNE2ROSfLpg27vm913qPOq6iNV9b6qekYnpZuIJjEDANAFy8gcsObQsqw+OurQb+Y3J3lya+3BqrooyTVJzvy6HVXtSbInSXbu3Lm+kgIAAMzM0yOzL8lpB90+NcndB2/QWvtCa+3B2fXrkmyrqpMO3VFr7YrW2u7W2u4dO3ZsotjjIS0DAED35gkyH0pyZlWdXlXHJrkkybUHb1BVT6pa+cpeVc+d7ffergsLAAAkFviYY2hZa+3Rqnpzkv+S5Ogk72qt3VFVb5g9fnmSVyZ5Y1U9muSLSS5pJoYAAEDHavZfw37mmSOzf7jYdYfcd/lB1y9Ncmm3RQMAAFjdXAtisnGyMgAAdE+QAQCAkbAg5lcJMgAAwOgIMltAYAYAoBOztT0s8SHI9K68ygAAoHOCDAAAjEQ7sH6MIT+CDAAAMDqCzBZoEjMAAJ2og/673ASZnnmRAQBA9wQZAAAYCevIfJUgAwAAjI4gswUEZgAAOmEdmQMEGQAAYHQEmZ5JywAAdMVAn68SZAAAgNERZLaAOTIAAHTDOjL7CTIAAMDoCDI9K3kZAAA6J8gAAMBItP1zFsxdEGS2QnN+CQAAurD/lLgG/QgyAADA+AgyfZOWAQCgc4IMAACMxP4JC2XqgiCzFczFAgCAbgkyAAAwGuYt7CfI9MxLDQAAuifIAADASJiy8FWCzBbwegMAgG4JMgAAMBazBTFNXxBkeldeZQAA0DlBBgAAxqJ93ZWlJchsBa8zAADolCADAACjYY7MfoJMz8rLDAAAOifIAAAAoyPIbIFmkgwAAB048L3S10tBBgAARmP/OjJmLwgyffMiAwCA7gkyAADA6AgyW6AZwwgAQAeaBTEPEGQAAIDREWR6Zo4MAADd8eVyP0EGAAAYHUEGAABGYv/MmDIJW5DZCl5mAADQLUGmZ2UcIwAAXdk/AdtXTEEGAAAYH0EGAABGos3mxuiQEWS2RDMZCwAAOiXI9Mw6MgAAdMeXy/0EGQAAYHQEGQAAYHQEmS1ghgwAAF04sCCmb5jzBZmqurCq7qyqvVX1llUer6p6x+zx26rqnO6LCgAAS84E7APWDDJVdXSSdyZ5SZKnJ3lNVT39kM1ekuTM2WVPkss6LicAAMAB8/TIPDfJ3tbaX7bWHknyW0lefsg2L09yVVtxQ5ITqurkjssKAACQJDlmjm1OSfLpg27vS/K8ObY5Jck9myrdRPzV3/193vR/37zoYgAAMHJf+vJX8itJnvy3709+5zP9PMnO85Ln/c/97LtD8wSZ1QbiHTq7aJ5tUlV7sjL0LDt37pzjqcfvBWftyN2f/2I+cc8XFl0UAAAm4C+2fXuec8xnks/c0c8TfONp/ey3Y/MEmX1JDv7XnJrk7g1sk9baFUmuSJLdu3cvxakWfuA7npIf+I6nLLoYAABMxgsXXYBBmGeOzIeSnFlVp1fVsUkuSXLtIdtcm+T7ZmcvOzfJ/a01w8oAAIBerNkj01p7tKrenOS/JDk6ybtaa3dU1Rtmj1+e5LokFyXZm+ShJK/vr8gAAMCym2doWVpr12UlrBx83+UHXW9J3tRt0QAAAFY314KYAAAAQyLIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAoyPIAAAAo1OttcU8cdXnkvz1Qp58dScl+btFF4LOaM9p0Z7Toj2nR5tOi/aclrG355NbaztWe2BhQWZoquqm1truRZeDbmjPadGe06I9p0ebTov2nJYpt6ehZQAAwOgIMgAAwOgIMl91xaILQKe057Roz2nRntOjTadFe07LZNvTHBkAAGB09MgAAACjs/RBpqourKo7q2pvVb1l0eXh8Krqrqr6aFXdWlU3ze77pqr6w6r677P//w8Hbf8vZ+16Z1X9k4Pu/x9n+9lbVe+oqlrEv2fZVNW7quqzVXX7Qfd11n5V9biq+u3Z/R+sql1b+g9cQodp039dVX8zO05vraqLDnpMmw5YVZ1WVX9SVR+vqjuq6odm9ztOR+gI7ekYHaGq2l5VN1bVR2bt+dOz+5f7+GytLe0lydFJPpnkKUmOTfKRJE9fdLlcDttedyU56ZD7fj7JW2bX35Lk386uP33Wno9LcvqsnY+ePXZjkvOSVJL3JXnJov9ty3BJ8p1Jzklyex/tl+QHk1w+u35Jkt9e9L956pfDtOm/TvJjq2yrTQd+SXJyknNm149L8t9m7eY4HeHlCO3pGB3hZVb3T5hd35bkg0nOXfbjc9l7ZJ6bZG9r7S9ba48k+a0kL19wmViflye5cnb9yiT/7KD7f6u19qXW2l8l2ZvkuVV1cpLjW2vXt5Uj9aqD/oYetdb+PMl9h9zdZfsdvK//mOTFetv6dZg2PRxtOnCttXtaazfPrj+Q5ONJTonjdJSO0J6Hoz0HrK14cHZz2+zSsuTH57IHmVOSfPqg2/ty5IOcxWpJ3l9VH66qPbP7nthauydZedNO8g9n9x+ubU+ZXT/0fhajy/Y78DettUeT3J/kxN5KzpG8uapumw092z/MQZuOyGxIybOz8quv43TkDmnPxDE6SlV1dFXdmuSzSf6wtbb0x+eyB5nVUqbTuA3X+a21c5K8JMmbquo7j7Dt4dpWm4/DRtpP2w7DZUmemuRZSe5J8u9m92vTkaiqJyR5T5Ifbq194UibrnKfNh2YVdrTMTpSrbXHWmvPSnJqVnpXvvUImy9Fey57kNmX5LSDbp+a5O4FlYU1tNbunv3/s0nem5WhgZ+ZdZNm9v/PzjY/XNvum10/9H4Wo8v2O/A3VXVMkm/M/MOe6Ehr7TOzD9uvJPnlrByniTYdharalpUvvb/ZWvvd2d2O05FarT0do+PXWvt8kj9NcmGW/Phc9iDzoSRnVtXpVXVsViY2XbvgMrGKqnp8VR23/3qS705ye1ba63WzzV6X5Pdm169NcsnsDBynJzkzyY2zbtcHqurc2bjP7zvob9h6Xbbfwft6ZZI/no3/ZQvt/0Cd+edZOU4TbTp4s/r/1SQfb629/aCHHKcjdLj2dIyOU1XtqKoTZte/IckFST6RZT8+F322gUVfklyUlTN5fDLJv1p0eVwO205PycrZNz6S5I79bZWVsZv/Ncl/n/3/mw76m381a9c7c9CZyZLszsob9yeTXJrZwrAuvbfh1VkZxvDlrPzq8y+6bL8k25P8TlYmNN6Y5CmL/jdP/XKYNv31JB9NcltWPhRP1qbjuCR5flaGkdyW5NbZ5SLH6TgvR2hPx+gIL0nOTnLLrN1uT/KTs/uX+vjcX3AAAIDRWPahZQAAwAgJMgAAwOgIMgAAwOgIMgAAwOgIMgAAwOgIMgAAwOgIMgAAwOgIMgAAwOj8/x/tmcBfCyLxAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1008x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### shift positions\n",
    "def plot_overlap_check(gap_number):\n",
    "    \n",
    "    # Reset ORIG_POS to zero for plotting\n",
    "    shift = min(nanopore_reads['ORIG_POS'])\n",
    "    nanopore_reads['ORIG_POS_plot'] = nanopore_reads['ORIG_POS']-shift\n",
    "    nanopore_reads['ORIG_END_POS_plot'] = nanopore_reads['ORIG_END_POS']-shift\n",
    "\n",
    "    # Create template\n",
    "    pre_overlap  = np.zeros(len(range(min(nanopore_reads['ORIG_POS_plot']),max(nanopore_reads['ORIG_END_POS_plot']))))\n",
    "    post_overlap = np.zeros(len(range(min(nanopore_reads['ORIG_POS_plot']),max(nanopore_reads['ORIG_END_POS_plot']))))\n",
    "\n",
    "#     post_overlap = np.zeros(len(range(min(nanopore_reads['POS'])          ,max(nanopore_reads['END_POS']          ))))\n",
    "\n",
    "    # Plot\n",
    "    selected=confirmed_gaps[gap_number]\n",
    "    padding=1000\n",
    " \n",
    "    # Center post_overlap reads at beginning of gap\n",
    "    key_read = nanopore_reads.loc[nanopore_reads['ORIG_POS']>selected[0]].iloc[0]\n",
    "    \n",
    "    # Calculate shift in first read\n",
    "    shift2 = key_read['ORIG_POS_plot']- key_read['POS']\n",
    "    \n",
    "    # Shift all reads to align in plot\n",
    "    nanopore_reads['POS_plot']     = nanopore_reads['ORIG_POS_plot'] + shift2\n",
    "    nanopore_reads['END_POS_plot'] = nanopore_reads['ORIG_END_POS_plot'] + shift2\n",
    "\n",
    "#     nanopore_reads.loc[nanopore_reads['ORIG_POS']-padding>selected[0],'END_POS_plot'] = nanopore_reads.loc[nanopore_reads['ORIG_POS']-padding>selected[0],'ORIG_END_POS_plot'] + shift2\n",
    "    \n",
    "# nanopore_reads.loc[nanopore_reads['ID']>=]\n",
    "#     print(key_read)\n",
    "    \n",
    "#     print(shift2)\n",
    "#     print(selected[1]-selected[0]+1)\n",
    "    \n",
    "    # Populate template\n",
    "#     count = 0\n",
    "    for index,read in nanopore_reads.iterrows():\n",
    "    #     print(read['ORIG_POS_plot'])\n",
    "    #     print(read['POS'])\n",
    "        pre_overlap [read['ORIG_POS_plot']:read['ORIG_END_POS_plot']] += 1\n",
    "        post_overlap[read['POS_plot'] :read['END_POS_plot'] ] += 1\n",
    "#         count +=1\n",
    "\n",
    "    # print(count)\n",
    "  \n",
    "    pre_overlap_plot = pre_overlap[selected[0]-padding-shift:selected[1]+padding-shift]\n",
    "    post_overlap_plot = post_overlap[selected[0]-padding-shift:selected[1]+padding-shift]\n",
    "\n",
    "    fig,ax = plt.subplots(figsize=(14,10))\n",
    "    ax=plt.plot(pre_overlap_plot, label=\"pre\")\n",
    "    ax2=plt.plot(post_overlap_plot, label=\"post\")\n",
    "    plt.legend()\n",
    "\n",
    "gap_number=0\n",
    "plot_overlap_check(gap_number)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adjust reference_genome based on removed indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 29 gaps = 114,721 basepairs.\n",
      "Reference genome length before collapsing gaps: 4,080,269 basepairs.\n",
      "Final genome length should be: 3,965,548\n",
      "Final genome length after collapsing gaps: 3,965,548 basepairs.\n"
     ]
    }
   ],
   "source": [
    "# Reset confirmed_gaps to start at 0\n",
    "shift = min(nanopore_reads['ORIG_POS'])\n",
    "orig_confirmed_gaps = [(gap[0]-shift, gap[1]-shift) for gap in confirmed_gaps]\n",
    "\n",
    "indices_to_remove = []\n",
    "for gap in orig_confirmed_gaps:\n",
    "    for i in range(gap[0],gap[1]+1):\n",
    "        indices_to_remove.append(i)\n",
    "\n",
    "print(f\"Removing {len(confirmed_gaps)} gaps = {len(indices_to_remove):,} basepairs.\")\n",
    "print(f\"Reference genome length before collapsing gaps: {len(ref_genome):,} basepairs.\")\n",
    "print(f\"Final genome length should be: {len(ref_genome)-len(indices_to_remove):,}\")\n",
    "\n",
    "final_genome = ''\n",
    "segment_start = 0\n",
    "\n",
    "for gap in orig_confirmed_gaps:\n",
    "\n",
    "    gap_start = gap[0]\n",
    "    gap_end = gap[1]\n",
    "    \n",
    "    # Cut segment until next gap\n",
    "    new_segment = ref_genome[segment_start:gap_start]\n",
    "    \n",
    "    # Append to final genome\n",
    "    final_genome = final_genome + new_segment\n",
    "    \n",
    "    # Start of next segment\n",
    "    segment_start = gap_end + 1\n",
    "    \n",
    "# Add last segment\n",
    "new_segment = ref_genome[segment_start:]\n",
    "final_genome = final_genome + new_segment\n",
    "\n",
    "print(f\"Final genome length after collapsing gaps: {len(final_genome):,} basepairs.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Reference Genome\n",
    "initial_position = min(nanopore_reads['ORIG_POS'])\n",
    "final_genome_dict = {'initial_position':initial_position, 'reference_genome': final_genome}\n",
    "with open(f'{datadir}/{run_number}_{chrom_dis}_reference_genome.json', 'w') as f:\n",
    "    json.dump(final_genome_dict,f)\n",
    "\n",
    "# Save Nanopore Reads\n",
    "nanopore_reads.to_csv(f'{datadir}/{run_number}_{chrom_dis}_clean.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EXTRA CODE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Gaps V2\n",
    "\n",
    "This version identifies gaps in reads by running two loops. It was discarded because the current version does the same job with just one loop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# confirmed_gaps2 = []\n",
    "# for index in nanopore_reads.index:\n",
    "#     try:\n",
    "#         current_read_end = nanopore_reads.loc[index,'END_POS']\n",
    "        \n",
    "#         gaps = []\n",
    "#         for index2 in nanopore_reads.index:\n",
    "#             try:\n",
    "#                 if index == index2: # it's the same read\n",
    "#                     continue\n",
    "\n",
    "# #                 next_read_start = nanopore_reads.loc[index+1,'POS']\n",
    "#                 next_read_start = nanopore_reads.loc[index2,'POS']\n",
    "\n",
    "#                 delta = next_read_start-current_read_end\n",
    "\n",
    "\n",
    "#                 if delta < 0: # There is overlap or it's a previous read\n",
    "#                     continue \n",
    "#                 elif delta==0: # There is a single nucleotide overlap\n",
    "#                     continue\n",
    "#                 elif delta > 0: # There is a gap\n",
    "#                     gaps.append((current_read_end+1,next_read_start-1)) # range between the values\n",
    "#             except:\n",
    "#                 pass\n",
    "            \n",
    "                \n",
    "                \n",
    "#         minimum_gap = min(gaps, key = lambda x: x[1]-x[0])\n",
    "#         confirmed_gaps2.append(minimum_gap)\n",
    "        \n",
    "#     except:\n",
    "#         pass\n",
    "\n",
    "# print(f\"There are {len(confirmed_gaps2)} gaps\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collapse gaps\n",
    "\n",
    "This older version takes longer to delete gaps since it iterates over a list instead of using the indexes of strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ref_genome_list = list(ref_genome)\n",
    "# for index in sorted(indices_to_remove, reverse=True):\n",
    "#     del ref_genome_list[index]\n",
    "# ref_genome = ''.join(ref_genome_list)\n",
    "\n",
    "# print(f\"Reference genome length after collapsing gaps: {len(ref_genome):,}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
