{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"\"\n",
      "[1] \"------------------------------------------------------------------------------------------\"\n",
      "[1] \"16 - DISEASE DIAGNOSTIC (16_disease_diagnostic.R)\"\n",
      "[1] \"Run: /home/fer/.local/share/jupyter/runtime/kernel-4006da55-029b-4016-8e60-df46edccae0c.json, disease: NA\"\n",
      "[1] \"\"\n",
      "[1] \"\"\n",
      "[1] \"Loading modules.\"\n",
      "[1] \"Loading data.\"\n"
     ]
    }
   ],
   "source": [
    "# 16 - Disease Diagnostic\n",
    "#\n",
    "# This script performs disease diagnostic. Reads bam files of two persons and extracts relevant reads to analyze.\n",
    "#\n",
    "# INPUTS:\n",
    "#   run_number\n",
    "#   disease\n",
    "#   BAM files\n",
    "# \n",
    "# OUTPUTS:\n",
    "#   relevant reads with the SNP\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "print('')\n",
    "print('------------------------------------------------------------------------------------------')\n",
    "print('16 - DISEASE DIAGNOSTIC (16_disease_diagnostic.R)')\n",
    "print(paste0(\"Run: \", args[1], \", disease: \", args[2]))\n",
    "print('')\n",
    "print('')\n",
    "\n",
    "# Set variables\n",
    "# -----------------------------------------------------------------------------\n",
    "run_num = 3\n",
    "dis = 'sca'\n",
    "\n",
    "# Load Modules\n",
    "# -----------------------------------------------------------------------------\n",
    "print('Loading modules.')\n",
    "suppressMessages(library(tidyverse, quietly=TRUE))\n",
    "suppressMessages(library(Rsamtools, quietly=TRUE))\n",
    "suppressMessages(library(dplyr, quietly=TRUE))\n",
    "\n",
    "# Set Parameters\n",
    "# -----------------------------------------------------------------------------\n",
    "# print('Setting parameters.')\n",
    "\n",
    "run_number=paste0(\"run\",run_num)\n",
    "\n",
    "dis_params = data.frame('disease' = c('sca', 'cf', 'sma1', 'sma2', 'thal1', 'thal2', 'thal3', 'pompe'),\n",
    "                    'chr' = c('chr11','chr7','chr5', 'chr5', 'chr16', 'chr16', 'chr11', 'chr17'),\n",
    "                    'start' = c(5227002, 117559590, 70924941, 70049523, 176680, 172876, 5225464, 25000000),\n",
    "                    'end'   = c(5227002, 117559590, 70953015, 70077595, 177522, 173710, 5227071, 25000000))\n",
    "\n",
    "chr       = dis_params %>% filter(disease == dis) %>% select(chr)   %>% pull\n",
    "snp_start = dis_params %>% filter(disease == dis) %>% select(start) %>% pull\n",
    "snp_end   = dis_params %>% filter(disease == dis) %>% select(end)   %>% pull\n",
    "\n",
    "datadir = paste0('/mnt/aretian/genomics/nanopore/',run_number)\n",
    "\n",
    "# Load data\n",
    "# -----------------------------------------------------------------------------\n",
    "print('Loading data.')\n",
    "\n",
    "person0bam <- BamFile(\n",
    "                file  = paste0(datadir,'/strspy/',dis,'/input/',run_number,'_',chr,'_',dis,'_person0.bam'    ), \n",
    "                index = paste0(datadir,'/strspy/',dis,'/input/',run_number,'_',chr,'_',dis,'_person0.bam.bai'))\n",
    "\n",
    "person1bam <- BamFile(\n",
    "                file  = paste0(datadir,'/strspy/',dis,'/input/',run_number,'_',chr,'_',dis,'_person1.bam'    ), \n",
    "                index = paste0(datadir,'/strspy/',dis,'/input/',run_number,'_',chr,'_',dis,'_person1.bam.bai'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Extracting reads containing the relevant SNP.\"\n",
      "[1] \"There are no reads in location chr11:5227002-5227002\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in value[[3L]](cond): failed to open BamFile: file(s) do not exist:\n  ‘/mnt/aretian/genomics/nanopore/run3/strspy/sca/input/run3_chr11_sca_person1.bam’\n",
     "output_type": "error",
     "traceback": [
      "Error in value[[3L]](cond): failed to open BamFile: file(s) do not exist:\n  ‘/mnt/aretian/genomics/nanopore/run3/strspy/sca/input/run3_chr11_sca_person1.bam’\nTraceback:\n",
      "1. scanBam(person1bam, param = params)",
      "2. scanBam(person1bam, param = params)",
      "3. open(file)",
      "4. open.BamFile(file)",
      "5. tryCatch({\n .     .io_check_exists(path(con))\n .     index <- sub(\"\\\\.bai$\", \"\", index(con, asNA = FALSE))\n .     con$.extptr <- .Call(.bamfile_open, path(con), index, \"rb\")\n . }, error = function(err) {\n .     stop(\"failed to open BamFile: \", conditionMessage(err))\n . })",
      "6. tryCatchList(expr, classes, parentenv, handlers)",
      "7. tryCatchOne(expr, names, parentenv, handlers[[1L]])",
      "8. value[[3L]](cond)",
      "9. stop(\"failed to open BamFile: \", conditionMessage(err))"
     ]
    }
   ],
   "source": [
    "# Diagnostic and save table\n",
    "# -----------------------------------------------------------------------------\n",
    "print('Extracting reads containing the relevant SNP.')\n",
    "\n",
    "# gr <- GRanges(seqnames = chr,\n",
    "#               ranges = IRanges(start = snp_start, end = snp_end+1000))\n",
    "gr <- GRanges(seqnames = 'chr11',\n",
    "              ranges = IRanges(start = 5227002, end = 5228002))\n",
    "params <- ScanBamParam(which = gr, what = scanBamWhat())\n",
    "\n",
    "# Person 0\n",
    "aln <- scanBam(person0bam, param = params)\n",
    "reads <- data.frame('startpos' = aln[[1]]$pos, \n",
    "                    'read'     = aln[[1]]$seq)\n",
    "tryCatch({ \n",
    "    reads2 <- reads                                                    %>%\n",
    "    mutate(length         = nchar(read)                   ,\n",
    "           snp_start_0    = snp_start - startpos          ,\n",
    "           snp_end_0      = snp_end   - startpos          ,\n",
    "           left_padding   = str_sub(read,0,snp_start_0   ),\n",
    "           right_padding  = str_sub(read,  snp_end_0  +2 ),\n",
    "           segment        = str_sub(read,  snp_start_0+1  , snp_end_0+1),\n",
    "           snp_start      = snp_start                     ,\n",
    "           snp_end        = snp_end                       )            %>%\n",
    "    select(snp_start, snp_end, left_padding,  segment, right_padding)  %>%\n",
    "    sample_n(10)\n",
    "    \n",
    "    write.csv(reads2, paste0(datadir,'/',run_number,'_',chr,'_',dis,'person0_diagnostic_reads.csv'), row.names = FALSE)\n",
    "    write.csv(reads2, paste0('disease_diagnostic/',run_number,'_',chr,'_',dis,'person0_diagnostic_reads.csv'), row.names = FALSE)\n",
    "    print(paste0('Saved: ',datadir,'/',run_number,'_',chr,'_',dis,'person0_diagnostic_reads.csv'))\n",
    "    print('Saved copies of each csv in disease_diagnostic/ folder in homedir.')    \n",
    "    }, \n",
    "    error=function(cond){\n",
    "        print(paste0('There are no reads in location ', chr,':',snp_start,'-',snp_end))\n",
    "    })\n",
    "\n",
    "\n",
    "\n",
    "# Person 1\n",
    "aln <- scanBam(person1bam, param = params)\n",
    "reads <- data.frame('startpos' = aln[[1]]$pos, \n",
    "                    'read' = aln[[1]]$seq)\n",
    "\n",
    "tryCatch({ \n",
    "    reads2 <- reads                                                    %>%\n",
    "    mutate(length         = nchar(read)                   ,\n",
    "           snp_start_0    = snp_start - startpos          ,\n",
    "           snp_end_0      = snp_end   - startpos          ,\n",
    "           left_padding   = str_sub(read,0,snp_start_0   ),\n",
    "           right_padding  = str_sub(read,  snp_end_0  +2 ),\n",
    "           segment        = str_sub(read,  snp_start_0+1  , snp_end_0+1),\n",
    "           snp_start      = snp_start                     ,\n",
    "           snp_end        = snp_end                       )            %>%\n",
    "    select(snp_start, snp_end, left_padding,  segment, right_padding)  %>%\n",
    "    sample_n(10)\n",
    "    \n",
    "    write.csv(reads2, paste0(datadir,'/',run_number,'_',chr,'_',dis,'person1_diagnostic_reads.csv'), row.names = FALSE)\n",
    "    write.csv(reads2, paste0('disease_diagnostic/',run_number,'_',chr,'_',dis,'person1_diagnostic_reads.csv'), row.names = FALSE)\n",
    "\n",
    "    print(paste0('Saved: ',datadir,'/',run_number,'_',chr,'_',dis,'person1_diagnostic_reads.csv'))\n",
    "    print('Saved copies of each csv in disease_diagnostic/ folder in homedir.')\n",
    "    }, \n",
    "    error=function(cond){\n",
    "        print(paste0('There are no reads in location ', chr,':',snp_start,'-',snp_end))\n",
    "    })\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
