{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering\n",
    "\n",
    "This script performs the clustering algorithm. It contains the following steps:\n",
    "\n",
    "- Padding reads with the reference genome to fill empty spaces\n",
    "- Sliding a window over the region of interest to identify reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "encoder = OneHotEncoder()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Nanopore reads\n",
    "nanopore_reads = pd.read_csv('../data/processed/bioliquid_chr17_pompe_reads_nogaps.csv')\n",
    "\n",
    "# Import reference genome for the location of interest without gaps\n",
    "with open('../data/processed/reference_genome_chr17_nogaps.json', 'r') as f:\n",
    "    ref_genome_json = json.load(f)\n",
    "    \n",
    "ref_genome = ref_genome_json['reference_genome']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define hyper-parameters\n",
    "window_width = 5000 # 5k BP\n",
    "jump_width = 1000 # 1k BP sliding window (ideally jump_width=1)\n",
    "\n",
    "# check if the start of a sequence is in range\n",
    "def check_sliding_window(read):\n",
    "    read_filter = False\n",
    "    window = set(list(range(left_bound,right_bound)))\n",
    "    read = set(list(range(read['POS'],read['END_POS'])))\n",
    "    if len(window & read)>0:\n",
    "        read_filter = True\n",
    "    return read_filter\n",
    "\n",
    "# Input: read --> Series\n",
    "# Output: nucleotide_sequence_trimmed: String\n",
    "\n",
    "def left_pad_read(read):\n",
    "    nucleotide_sequence = list(read['SEQ']) # turn sequence string into a list\n",
    "    genome_sequence = list(window_ref_genome)\n",
    "    \n",
    "    start = read['POS']-left_bound\n",
    "\n",
    "    if start == 0:\n",
    "        return read['SEQ'] # read starts on the start of the window\n",
    "    \n",
    "    elif start > 0: # sequence starts to the right of the bound\n",
    "#         print('HERE1')\n",
    "#         genome_fill = window_ref_genome[:start] # select portion of reference genome to pad the left with\n",
    "        genome_fill = genome_sequence[:start] # select portion of reference genome to pad the left with\n",
    "#         print('genome_fill type:', type(genome_fill))\n",
    "#         print('nucleotide_sequence type:', type(nucleotide_sequence))\n",
    "        nucleotide_sequence = genome_fill+nucleotide_sequence\n",
    "        nucleotide_sequence_trimmed = ''.join(nucleotide_sequence) # keep entire sequence\n",
    "        return nucleotide_sequence_trimmed\n",
    "    \n",
    "    elif start < 0: # sequence starts to the left of the bound\n",
    "#         print('HERE2')\n",
    "        nucleotide_sequence_trimmed = ''.join(nucleotide_sequence[np.abs(start):])\n",
    "        return nucleotide_sequence_trimmed\n",
    "\n",
    "# Input: read --> Series\n",
    "# Output: nucleotide_sequence_trimmed: String\n",
    "\n",
    "def right_pad_read(read):\n",
    "    nucleotide_sequence = list(read['left_padded']) # turn sequence string into a list\n",
    "    genome_sequence = list(window_ref_genome)\n",
    "\n",
    "    end = right_bound-read['END_POS']\n",
    "\n",
    "    if end == 0:\n",
    "        return read['left_padded'] # read end on the end of the window\n",
    "    \n",
    "    elif end > 0: # sequence ends to the left of the bound\n",
    "#         genome_fill = window_ref_genome[-end:] # select portion of reference genome to pad the right with\n",
    "        genome_fill = genome_sequence[-end:] # select portion of reference genome to pad the right with\n",
    "        nucleotide_sequence = nucleotide_sequence+genome_fill\n",
    "        nucleotide_sequence_trimmed = ''.join(nucleotide_sequence) # keep entire sequence\n",
    "        return nucleotide_sequence_trimmed\n",
    "    \n",
    "    elif end < 0: # sequence ends to the right of the bound\n",
    "        nucleotide_sequence_trimmed = ''.join(nucleotide_sequence[:end])\n",
    "        return nucleotide_sequence_trimmed\n",
    "\n",
    "# Should be updated\n",
    "NUCLEOTIDE_VOCABULARY = [\n",
    "    'A','C','G','T','X'\n",
    "]\n",
    "        \n",
    "# Not being used, instead, we use the sklearn one hot encoding\n",
    "def nucleotide_to_one_hot(nucleotide_sequence):\n",
    "    to_return = []\n",
    "    for char in nucleotide_sequence:\n",
    "        if char in NUCLEOTIDE_VOCABULARY:\n",
    "            to_append = np.zeros(len(NUCLEOTIDE_VOCABULARY))\n",
    "            to_append[NUCLEOTIDE_VOCABULARY.index(char)] = 1.\n",
    "            to_return.append(to_append)\n",
    "        else:\n",
    "            raise ValueError('Could not one-hot code character {}'.format(char))\n",
    "    return np.array(to_return)\n",
    "\n",
    "#nucleotide_to_one_hot('GTCATACX') # uncomment example to see what the encoding does\n",
    "\n",
    "# Input: read (Series)\n",
    "\n",
    "# Text file for results\n",
    "def write_results_to_file(read, results_file):\n",
    "    results_file.write(f\"{read['ID']},{read['kmeans_cls2']}\\n\")\n",
    "#     TODO: Add window number\n",
    "#     results_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Loop: padding and clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create results file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total range: 2,119,015\n",
      "Window width: 5000\n",
      "Jump width: 1000\n",
      "Number of iterations required: 2119\n",
      "Writing results to file: ../data/processed/chr17_read_clusters.txt\n",
      "Iterations:  0% 1111111111111111111111111111**"
     ]
    }
   ],
   "source": [
    "chromosomeID = 'chr17'\n",
    "results_file_path = f\"../data/processed/{chromosomeID}_read_clusters.txt\"\n",
    "results_file = open(results_file_path,\"w\")\n",
    "# results_file = open(results_file_path,\"a\")\n",
    "\n",
    "# window_ref_genome_start = 0\n",
    "iter = 1\n",
    "empty_count = 0\n",
    "print(f\"Total range: {max(nanopore_reads['END_POS']):,}\")\n",
    "print(f\"Window width: {window_width}\")\n",
    "print(f\"Jump width: {jump_width}\")\n",
    "total_jumps = round(max(nanopore_reads['END_POS'])/jump_width)\n",
    "print(f\"Number of iterations required: {total_jumps}\")\n",
    "print(f\"Writing results to file: {results_file_path}\")\n",
    "print(f\"Iterations: \", end=\"\")\n",
    "\n",
    "for left_bound in range(min(nanopore_reads['POS']),max(nanopore_reads['END_POS']),jump_width):\n",
    "    \n",
    "    progress = iter/total_jumps\n",
    "#     print(f\" {progress:.0f}% \", end=\"\")\n",
    "    if round(progress*1000)%50==0:\n",
    "        print(f\" {round(progress*100):.0f}% \", end=\"\")\n",
    "    iter+=1\n",
    "        \n",
    "    right_bound = left_bound+window_width\n",
    "    window_ref_genome = ref_genome[left_bound:left_bound+window_width]\n",
    "    \n",
    "    # Identify if each read is in the window: True/False\n",
    "#     print(\"Identifying if each read is in the window...\")\n",
    "    nanopore_reads['read_filter'] = nanopore_reads.apply(lambda x: check_sliding_window(x), axis=1)\n",
    "    \n",
    "    # Get them\n",
    "    window_reads = nanopore_reads.loc[nanopore_reads['read_filter']==True]\n",
    "    window_reads = window_reads.reset_index()\n",
    "    \n",
    "    # Display feedback to user every 100 empty windows\n",
    "#     if empty_count%100 == 0:\n",
    "#         print(empty_count)\n",
    "\n",
    "    if len(window_reads) == 0:\n",
    "        # There are no nanopore reads in this window (next window)\n",
    "#         window_ref_genome_start += jump_width # move the starting cutoff for the reference genome with the sliding window\n",
    "        empty_count += 1\n",
    "        print(f\"-\", end=\"\")\n",
    "        continue \n",
    "        \n",
    "    elif len(window_reads) == 1:\n",
    "        # There is only one read in this window (next window) - TODO: TALK WITH TEAM ABOUT THIS\n",
    "#         window_ref_genome_start += jump_width # move the starting cutoff for the reference genome with the sliding window\n",
    "        empty_count += 1\n",
    "        print(f\"1\", end=\"\")\n",
    "        continue\n",
    "        \n",
    "    ## If it gets here, there is more than one sequence overlapping in the window (continue algorithm to encoding and clustering steps) ##\n",
    "    \n",
    "#     print(\"Applying padding...\")\n",
    "    window_reads['left_padded'] = window_reads.apply(lambda x: left_pad_read(x), axis=1) # fill reference genome on the left of the read\n",
    "    window_reads['right_padded'] = window_reads.apply(lambda x: right_pad_read(x), axis=1) # fill reference genome on the right of the read\n",
    "    window_reads['final_padded_read'] = window_reads['right_padded']\n",
    "    \n",
    "    window_reads['FINAL_SEQ_LEN'] = window_reads['final_padded_read'].apply(lambda x: len(x)) # should always be 5k\n",
    "    \n",
    "    # not currently used. Should be used if we use the custom one-hot-encoding\n",
    "#     window_reads['one_hot_read_V1'] = window_reads['final_padded_read'].apply(lambda x: nucleotide_to_one_hot(x).flatten())  # apply one-hot encoding V1\n",
    "    \n",
    "    unique_reads = []\n",
    "    for index, read in window_reads.iterrows(): # TODO: TRY TO USE LAMBDA FUNCTION IF POSSIBLE\n",
    "        unique_reads.append(list(read['final_padded_read']))\n",
    "\n",
    "    # One hot encoder\n",
    "#     print(\"Running one hot encoder...\")\n",
    "    X_onehot = encoder.fit_transform(unique_reads).toarray()\n",
    "\n",
    "    # PCA\n",
    "#     print(\"Running PCA...\")\n",
    "    pca = PCA(n_components=2)\n",
    "    X_pca = pca.fit_transform(X_onehot)\n",
    "    \n",
    "    window_reads['PCA1'] = np.nan\n",
    "    window_reads['PCA2'] = np.nan\n",
    "    for ID in window_reads.index:\n",
    "        window_reads.loc[ID,'PCA1'] = X_pca[ID][0] # TODO: change rounding if desired\n",
    "        window_reads.loc[ID,'PCA2'] = X_pca[ID][1] # TODO: change rounding if desired\n",
    "        \n",
    "    # Standardizing the features\n",
    "    X = window_reads[['PCA1','PCA2']]\n",
    "    X = StandardScaler().fit_transform(X)\n",
    "\n",
    "    # Run Kmeans\n",
    "#     print(\"Running Kmeans...\")\n",
    "    model = KMeans(n_clusters=2, random_state=42)\n",
    "    cls2 = model.fit(X)\n",
    "    window_reads['kmeans_cls2'] = cls2.labels_\n",
    "    \n",
    "    # Write results to file\n",
    "    window_reads.apply(lambda x: write_results_to_file(x, results_file), axis=1) # write results to file\n",
    "    \n",
    "    print(f\"*\", end=\"\")\n",
    "    \n",
    "#     Break after N iterations\n",
    "#     if iter>30:\n",
    "#         break\n",
    "        \n",
    "results_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EXTRA CODE\n",
    "Optimize number of clusters. Not needed in this case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "\n",
    "# Standardizing the features\n",
    "X = test[['PCA1','PCA2']]\n",
    "X = StandardScaler().fit_transform(X)\n",
    "\n",
    "distortions = []\n",
    "\n",
    "for k in range(1, 5):\n",
    "    KMeans_model = KMeans(n_clusters=k, random_state=42)\n",
    "    KMeans_model.fit(X)\n",
    "    distortions.append(KMeans_model.inertia_)\n",
    "\n",
    "plt.plot(range(1, 5), distortions,  marker='o')\n",
    "plt.xlabel('Number of clusters (K)')\n",
    "plt.ylabel('Distortion')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "silhouette_scores = []\n",
    "for k in range(2, 5):\n",
    "    model = KMeans(n_clusters=k, random_state=42)\n",
    "    model.fit(X)\n",
    "    score = silhouette_score(X, model.labels_)\n",
    "    print(\"Silhouette Score for k = \", k, \"is\", score)\n",
    "    silhouette_scores.append(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(range(2, 5), silhouette_scores, marker='o')\n",
    "plt.xlabel('Number of clusters (K)')\n",
    "plt.ylabel('Silhouette score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from yellowbrick.cluster import KElbowVisualizer\n",
    "\n",
    "visualizer3 = KElbowVisualizer(KMeans(), k=(2,5))\n",
    "\n",
    "visualizer3.fit(X) # Fit the data to the visualizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = KMeans(n_clusters=2, random_state=42)\n",
    "cls2 = model.fit(X)\n",
    "test['kmeans_cls2'] = cls2.labels_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='PCA1', ylabel='PCA2'>"
      ]
     },
     "execution_count": 442,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAEGCAYAAABLgMOSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYsklEQVR4nO3de5CU9Z3v8fdX7kYQETDAkGUQvHAJoANicqyjRoNcFE0dE7O5WJvNejwnWnFdk2DMqaRSSZXGrWhZWhoTE/GsJ2gSb/FgjEk02VxUICKsXCKiWQcIsnpEkoBc/J4/pmGHsWGap6enZ5z3q6qr+/n9fs/T31936Yfn0s9EZiJJ0qE6rN4FSJK6JwNEklSIASJJKsQAkSQVYoBIkgrpXe8COtPQoUNzzJgx9S5DkrqVZcuW/UdmDmvb3qMCZMyYMSxdurTeZUhStxIRfyzX7iEsSVIhBogkqRADRJJUSI86ByJJu3btorm5mR07dtS7lC6nf//+NDQ00KdPn4rGGyCSepTm5mYGDhzImDFjiIh6l9NlZCavvvoqzc3NNDY2VrSOh7Ak9Sg7duzg6KOPNjzaiAiOPvroQ9ozM0Ak9TiGR3mH+rkYIJKkQgwQSVIhBoikHu2ll15i0qRJ9S6jKk888QTz5s07YP/y5cs59dRTmThxIu9973u55557OuR9vQpLkt7hDj/8cO666y7Gjx/Pxo0bOfnkk5k1axaDBw+uarvugUhSyfr165k2bRrXX389559/Pueeey6NjY3cfPPNfPOb32TatGnMnDmT1157DYAXXniBc845h5NPPpnTTjuNNWvWAPDjH/+YU045hWnTpnHWWWexefNmAL7yla/wqU99itNPP52xY8dy0003AfCXv/yFuXPnMmXKFCZNmnTQPYQlS5bwvve9jylTpjBjxgy2bdu2X/8vf/lLpk6dytSpU5k2bRrbtm3juOOOY/z48QCMHDmS4cOHs2XLluo/sMzsMY+TTz45JfVsq1at2m/5xRdfzIkTJ+aaNWty6tSp+cwzz+T3vve9PPbYY/ONN97IV155JQcNGpS33nprZmZeccUVecMNN2Rm5plnnpl/+MMfMjPzySefzDPOOCMzM1977bV86623MjPz29/+dl555ZWZmfnlL385Tz311NyxY0du2bIlhwwZkjt37swf/vCH+elPf3pfTa+//nrZ2t98881sbGzMp59+OjMzt27dmrt27crHH388586dm5mZ8+bNy1//+teZmblt27bctWvXftt46qmn8oQTTsg9e/ZU9PlkZgJLs8z/Uz2EJanH27JlC/Pnz+dHP/oREydOZPny5ZxxxhkMHDiQgQMHcuSRR3LuuecCMHnyZFasWMGf//xnfvvb33LhhRfu286bb74JtPxY8SMf+QibNm1i586d+/0wb+7cufTr149+/foxfPhwNm/ezOTJk7nqqqv4whe+wLx58zjttNPK1rl27VpGjBjB9OnTARg0aNDbxrz//e/nyiuv5GMf+xgf+tCHaGho2Ne3adMmPvGJT7Bw4UIOO6z6A1AewpLU4x155JGMHj2a3/zmN/va+vXrt+/1YYcdtm/5sMMOY/fu3bz11lsMHjyY5cuX73usXr0agMsvv5zLLruMlStX8q1vfWu/H+e13m6vXr3YvXs3xx13HMuWLWPy5MlcffXVfPWrXy1bZ2a2+1uNBQsW8J3vfIft27czc+bMfYfV3njjDebOncvXvvY1Zs6ceYifUHnugUjq8fr27csDDzzArFmzOOKIIypaZ9CgQTQ2NvKDH/yACy+8kMxkxYoVTJkyha1btzJq1CgAFi5c2O62Nm7cyJAhQ/j4xz/OEUccwZ133ll23AknnMDGjRtZsmQJ06dPZ9u2bQwYMGC/MS+88AKTJ09m8uTJ/O53v2PNmjWMHTuWCy64gE9+8pP77TFVyz0QSQLe9a538fDDD3PDDTewdevWita5++67ueOOO5gyZQoTJ07kwQcfBFpOll944YWcdtppDB06tN3trFy5khkzZjB16lS+/vWv86UvfansuL59+3LPPfdw+eWXM2XKFM4+++y33XrkxhtvZNKkSUyZMoUBAwYwe/Zs7r33Xn71q19x55137jvBvnz58ormeDDRcn6kZ2hqakr/IqHUs61evZoTTzyx3mV0WeU+n4hYlplNbce6ByJJKsRzIJLUBV1wwQW8+OKL+7Vdd911zJo1q04VvZ0BIkld0P3331/vEtrlISxJUiEGiCSpEANEklSIASJJ3chPfvITjj/+eMaNG8e1115b11rqGiARcU5ErI2IdRGxoEx/RMRNpf4VEXFSm/5eEfFMRDzceVVLUn3s2bOHz3zmMzzyyCOsWrWK73//+6xatapu9dTtKqyI6AXcApwNNANLIuKhzGz9acwGxpcepwC3lp73+iywGnj7HcUkqY4eeGYD1z+6lo2vb2fk4AF8btbxnD9tVFXbfPrppxk3bhxjx44F4KKLLuLBBx9kwoQJHVHyIavnHsgMYF1mrs/MncAiYH6bMfOBu0p3FH4SGBwRIwAiogGYC3ynM4uWpPY88MwGrr5vJRte304CG17fztX3reSBZzZUtd0NGzYwevTofcsNDQ1s2FDdNqtRzwAZBbzcarm51FbpmBuBzwNvHexNIuKSiFgaEUs75A+oSFI7rn90Ldt37dmvbfuuPVz/6Nqqtlvu1lPt3Z23luoZIOVm3fbTKTsmIuYBr2TmsvbeJDNvz8ymzGwaNmxYkTol6ZBsfH37IbVXqqGhgZdf/s9/Uzc3NzNy5MiqtlmNegZIMzC61XIDsLHCMe8HzouIl2g59HVmRPxL7UqVpMqNHDzgkNorNX36dJ5//nlefPFFdu7cyaJFizjvvPOq2mY16hkgS4DxEdEYEX2Bi4CH2ox5CPhk6WqsmcDWzNyUmVdnZkNmjimt94vM/HinVi9JB/C5WcczoE+v/doG9OnF52YdX9V2e/fuzc0338ysWbM48cQT+fCHP8zEiROr2mZV9dTrjTNzd0RcBjwK9AK+m5nPRcSlpf7bgMXAHGAd8Ffg7+pVryRVau/VVh19FRbAnDlzmDNnTtXb6Qh1vZliZi6mJSRat93W6nUCn2lnG08AT9SgPEkq7PxpozokMLoyf4kuSSrEAJHU4/Skv8R6KA71czFAJPUo/fv359VXXzVE2shMXn31Vfr371/xOv5BKUk9SkNDA83NzfjD4rfr378/DQ0NFY83QCT1KH369KGxsbHeZbwjeAhLklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklRIXQMkIs6JiLURsS4iFpTpj4i4qdS/IiJOKrWPjojHI2J1RDwXEZ/t/OolqWerW4BERC/gFmA2MAH4aERMaDNsNjC+9LgEuLXUvhv4p8w8EZgJfKbMupKkGqrnHsgMYF1mrs/MncAiYH6bMfOBu7LFk8DgiBiRmZsy8/cAmbkNWA2M6sziJamnq2eAjAJebrXczNtDoN0xETEGmAY81fElSpIOpJ4BEmXa8lDGRMQRwI+AKzLzjbJvEnFJRCyNiKVbtmwpXKwkaX/1DJBmYHSr5QZgY6VjIqIPLeFxd2bed6A3yczbM7MpM5uGDRvWIYVLkuobIEuA8RHRGBF9gYuAh9qMeQj4ZOlqrJnA1szcFBEB3AGszsxvdm7ZkiSA3vV648zcHRGXAY8CvYDvZuZzEXFpqf82YDEwB1gH/BX4u9Lq7wc+AayMiOWlti9m5uJOnIIk9WiR2fa0wztXU1NTLl26tN5lSFK3EhHLMrOpbbu/RJckFWKASJIKMUAkSYUYIJKkQgwQSVIhBogkqRADRJJUiAEiSSrEAJEkFWKASJIKMUAkSYUYIJKkQgwQSVIhBogkqRADRJJUiAEiSSrEAJEkFWKASJIKMUAkSYUYIJKkQgwQSVIhBogkqRADRJJUiAEiSSqkogCJiD5l2oZ2fDmSpO7ioAESEWdERDOwMSJ+GhFjWnX/tKaVSZK6tPb2QL4BzMrMYcDtwGMRMbPUFzWtTJLUpfVup79vZj4HkJk/jIjVwH0RsQDImlcnSeqy2guQXRHx7sz8E0BmPhcRHwAeBo6teXWSpC6rvUNYC4BjWjdkZjNwOnBtjWqSJHUDB90DycyfHaBrILCz48uRJHUXFf8OJCKGRsT/iIhfAU/QZs+kiIg4JyLWRsS60nmVtv0RETeV+ldExEmVrit1B8v++P+45fF1/OOiZ7h3ycuse2VbvUuSKnbQPZCIGAhcAPwtcBxwPzA2MxuqfeOI6AXcApwNNANLIuKhzFzVathsYHzpcQpwK3BKhetKXdqqjVv5/A+f5YUtfwHg/uUb+fRpjVx19vH079urztVJ7WtvD+QV4O+BrwPHZuY/0XGHrmYA6zJzfWbuBBYB89uMmQ/clS2eBAZHxIgK15W6tLWbt+0Lj73+9+/+yOpNb9SpIunQtBcgXwT60/Iv/6sjoiOvvBoFvNxqubnUVsmYStYFICIuiYilEbF0y5YtVRctdZTde95+JfyuPW+xO71CXt3DQQMkM2/IzFOA82j54eADwMiI+EJEHFfle5f7IWLb/3IONKaSdVsaM2/PzKbMbBo2bNghlijVzvhjjmDIu/ru1zZn8gjGD39XnSqSDk1FJ9FLh4q+npmTgenAkcAjVb53MzC61XIDsLHCMZWsK3VpU0cfxS1/O435U0Zy4oiBXH7mOP77aWMZfHi/epcmVaS9k+jjgGMy8zd72zJzZUQcBXy3yvdeAoyPiEZgA3ARLSfrW3sIuCwiFtFyEn1rZm6KiC0VrCt1eaceO5ST3nMUb+zYxbCB/etdjnRI2tsDuREod13hX4EbqnnjzNwNXAY8CqwG7i390v3SiLi0NGwxsB5YB3wb+J8HW7eaeqR66denl+GhbinyICfsIuLfMnPSAfpWlg5pdRtNTU25dOnSepchSd1KRCzLzKa27e3tgRzsn0UDqitJktSdtRcgSyLiH9o2RsTfA8tqU5IkqTto7268VwD3R8TH+M/AaAL60vILdUlSD9XezRQ3A++LiDOAvedC/m9m/qLmlUmSurT2LuPtD1wKjANWAneUroCSJPVw7Z0DWUjLIauVtNzY8J9rXpEkqVto7xzIhL2X6kbEHcDTtS9JktQdtLcHsmvvCw9dSZJaa28PZEpE7L23dAADSssBZGYOqml1kqQuq72rsPyrNpKksir+k7aSJLVmgEiSCjFAJEmFGCCSpEIMEElSIQaIJKkQA0SSVIgBIkkqxACRJBVigEiSCjFAJEmFGCCSpEIMEElSIQaIJKkQA0SSVIgBIkkqxACRJBVigEiSCjFAJEmFGCCSpELqEiARMSQiHouI50vPRx1g3DkRsTYi1kXEglbt10fEmohYERH3R8TgTitekgTUbw9kAfDzzBwP/Ly0vJ+I6AXcAswGJgAfjYgJpe7HgEmZ+V7gD8DVnVK1JGmfegXIfGBh6fVC4PwyY2YA6zJzfWbuBBaV1iMzf5qZu0vjngQaaluuJKmtegXIMZm5CaD0PLzMmFHAy62Wm0ttbX0KeKTDK5QkHVTvWm04In4GvLtM1zWVbqJMW7Z5j2uA3cDdB6njEuASgPe85z0VvrUkqT01C5DMPOtAfRGxOSJGZOamiBgBvFJmWDMwutVyA7Cx1TYuBuYBH8jM5AAy83bgdoCmpqYDjpMkHZp6HcJ6CLi49Ppi4MEyY5YA4yOiMSL6AheV1iMizgG+AJyXmX/thHolSW3UK0CuBc6OiOeBs0vLRMTIiFgMUDpJfhnwKLAauDcznyutfzMwEHgsIpZHxG2dPQFJ6ulqdgjrYDLzVeADZdo3AnNaLS8GFpcZN66mBUqS2uUv0SVJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVYoBIkgoxQCRJhRggkqRCDBBJUiEGiCSpEANEklSIASJJKsQAkSQVUpcAiYghEfFYRDxfej7qAOPOiYi1EbEuIhaU6b8qIjIihta+aklSa/XaA1kA/DwzxwM/Ly3vJyJ6AbcAs4EJwEcjYkKr/tHA2cC/d0rFkqT91CtA5gMLS68XAueXGTMDWJeZ6zNzJ7CotN5eNwCfB7KGdUqSDqBeAXJMZm4CKD0PLzNmFPByq+XmUhsRcR6wITOfbe+NIuKSiFgaEUu3bNlSfeWSJAB612rDEfEz4N1luq6pdBNl2jIiDi9t44OVbCQzbwduB2hqanJvRZI6SM0CJDPPOlBfRGyOiBGZuSkiRgCvlBnWDIxutdwAbASOBRqBZyNib/vvI2JGZv6pwyYgSTqoeh3Cegi4uPT6YuDBMmOWAOMjojEi+gIXAQ9l5srMHJ6ZYzJzDC1Bc5LhIUmdq14Bci1wdkQ8T8uVVNcCRMTIiFgMkJm7gcuAR4HVwL2Z+Vyd6pUktVGzQ1gHk5mvAh8o074RmNNqeTGwuJ1tjeno+iRJ7fOX6JKkQgwQSVIhBogkqRADRJJUiAEiSSrEAJEkFWKASJIKMUAkSYUYIJKkQgwQSVIhBogkqRADRJJUiAEiSSrEAJEkFWKASJIKMUAkSYUYIJKkQgwQSVIhBogkqRADRJJUiAEiSSrEAJEkFWKASJIKMUAkSYVEZta7hk4TEVuAP9a7jgKGAv9R7yI6UU+bLzjnnqK7zvlvMnNY28YeFSDdVUQszcymetfRWXrafME59xTvtDl7CEuSVIgBIkkqxADpHm6vdwGdrKfNF5xzT/GOmrPnQCRJhbgHIkkqxACRJBVigHQBETEkIh6LiOdLz0cdYNw5EbE2ItZFxIIy/VdFREbE0NpXXZ1q5xwR10fEmohYERH3R8TgTiv+EFXwvUVE3FTqXxERJ1W6bldVdM4RMToiHo+I1RHxXER8tvOrL6aa77nU3ysinomIhzuv6iplpo86P4BvAAtKrxcA15UZ0wt4ARgL9AWeBSa06h8NPErLDyWH1ntOtZ4z8EGgd+n1deXW7wqP9r630pg5wCNAADOBpypdtys+qpzzCOCk0uuBwB/e6XNu1X8l8H+Ah+s9n0of7oF0DfOBhaXXC4Hzy4yZAazLzPWZuRNYVFpvrxuAzwPd5aqIquacmT/NzN2lcU8CDbUtt7D2vjdKy3dliyeBwRExosJ1u6LCc87MTZn5e4DM3AasBkZ1ZvEFVfM9ExENwFzgO51ZdLUMkK7hmMzcBFB6Hl5mzCjg5VbLzaU2IuI8YENmPlvrQjtQVXNu41O0/MuuK6pkDgcaU+n8u5pq5rxPRIwBpgFPdXyJHa7aOd9Iyz8A36pRfTXRu94F9BQR8TPg3WW6rql0E2XaMiIOL23jg0Vrq5VazbnNe1wD7AbuPrTqOk27czjImErW7YqqmXNLZ8QRwI+AKzLzjQ6srVYKzzki5gGvZOayiDi9owurJQOkk2TmWQfqi4jNe3ffS7u0r5QZ1kzLeY69GoCNwLFAI/BsROxt/31EzMjMP3XYBAqo4Zz3buNiYB7wgSwdRO6CDjqHdsb0rWDdrqiaORMRfWgJj7sz874a1tmRqpnzfwPOi4g5QH9gUET8S2Z+vIb1dox6n4TxkQDXs/8J5W+UGdMbWE9LWOw9STexzLiX6B4n0auaM3AOsAoYVu+5tDPPdr83Wo59tz65+vShfOdd7VHlnAO4C7ix3vPorDm3GXM63egket0L8JEARwM/B54vPQ8ptY8EFrcaN4eWq1JeAK45wLa6S4BUNWdgHS3Hk5eXHrfVe04Hmevb5gBcClxaeh3ALaX+lUDToXznXfFRdM7Af6Hl0M+KVt/tnHrPp9bfc6ttdKsA8VYmkqRCvApLklSIASJJKsQAkSQVYoBIkgoxQCRJhRggUg1ExJ6IWB4R/xYRPyjdMYCIeHdELIqIFyJiVUQsjojjWq33jxGxIyKObNV2dOkOtX+OiJvrMR+pHANEqo3tmTk1MycBO4FLo+VWAfcDT2TmsZk5AfgicEyr9T4KLAEuaNW2A/hfwFWdU7pUGQNEqr1/BcYBZwC7MvO2vR2ZuTwz/xUgIo4FjgC+REuQ7B3zl8z8NS1BInUZBohUQxHRG5hNyy+PJwHLDjL8o8D3aQmc4yOi3B2KpS7DAJFqY0BELAeWAv8O3FHBOhcBizLzLeA+4MLalSdVz7vxSrWxPTOntm6IiOdoufPq20TEe4HxwGOluyr3peXmfLfUtkypOPdApM7zC6BfRPzD3oaImB4R/5WWw1dfycwxpcdIYFRE/E29ipXa480UpRqIiD9n5hFl2kfS8tfnTqblpPhLwBXAT4HZmbmm1dhvApsz87qIeAkYRMueyevABzNzVU0nIbXDAJEkFeIhLElSIQaIJKkQA0SSVIgBIkkqxACRJBVigEiSCjFAJEmF/H/3hWmiRXiUtQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.scatterplot(data=window_reads,x='PCA1',y='PCA2',hue='kmeans_cls2')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
